tls/s2n_fingerprint.c

```diff
@@ -19,6 +19,164 @@
 #include "utils/s2n_mem.h"
 #include "utils/s2n_safety.h"
 
+static S2N_RESULT s2n_fingerprint_init(struct s2n_fingerprint *fingerprint,
+        s2n_fingerprint_type type)
+{
+    RESULT_ENSURE_REF(fingerprint);
+
+    switch (type) {
+        case S2N_FINGERPRINT_JA3:
+            fingerprint->method = &ja3_fingerprint;
+            break;
+        default:
+            RESULT_BAIL(S2N_ERR_INVALID_ARGUMENT);
+    }
+
+    const struct s2n_fingerprint_method *method = fingerprint->method;
+    RESULT_ENSURE_REF(method);
+    RESULT_GUARD_POSIX(s2n_hash_new(&fingerprint->hash));
+    s2n_hash_allow_md5_for_fips(&fingerprint->hash);
+    RESULT_GUARD_POSIX(s2n_hash_init(&fingerprint->hash, method->hash));
+    return S2N_RESULT_OK;
+}
+
+struct s2n_fingerprint *s2n_fingerprint_new(s2n_fingerprint_type type)
+{
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_fingerprint)));
+    PTR_GUARD_POSIX(s2n_blob_zero(&mem));
+    struct s2n_fingerprint *fingerprint = (struct s2n_fingerprint *) (void *) mem.data;
+    PTR_ENSURE_REF(fingerprint);
+    PTR_GUARD_RESULT(s2n_fingerprint_init(fingerprint, type));
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return fingerprint;
+}
+
+static S2N_CLEANUP_RESULT s2n_fingerprint_free_fields(struct s2n_fingerprint *fingerprint)
+{
+    if (!fingerprint) {
+        return S2N_RESULT_OK;
+    }
+    RESULT_GUARD_POSIX(s2n_hash_free(&fingerprint->hash));
+    return S2N_RESULT_OK;
+}
+
+int s2n_fingerprint_free(struct s2n_fingerprint **fingerprint_ptr)
+{
+    if (!fingerprint_ptr) {
+        return S2N_SUCCESS;
+    }
+    POSIX_GUARD_RESULT(s2n_fingerprint_free_fields(*fingerprint_ptr));
+    POSIX_GUARD(s2n_free_object((uint8_t **) (void **) fingerprint_ptr,
+            sizeof(struct s2n_fingerprint)));
+    return S2N_SUCCESS;
+}
+
+int s2n_fingerprint_wipe(struct s2n_fingerprint *fingerprint)
+{
+    POSIX_ENSURE(fingerprint, S2N_ERR_INVALID_ARGUMENT);
+    fingerprint->client_hello = NULL;
+    fingerprint->raw_size = 0;
+    return S2N_SUCCESS;
+}
+
+int s2n_fingerprint_set_client_hello(struct s2n_fingerprint *fingerprint, struct s2n_client_hello *ch)
+{
+    POSIX_ENSURE(fingerprint, S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(ch, S2N_ERR_INVALID_ARGUMENT);
+    POSIX_ENSURE(!ch->sslv2, S2N_ERR_PROTOCOL_VERSION_UNSUPPORTED);
+    POSIX_GUARD(s2n_fingerprint_wipe(fingerprint));
```

Comment:
Yes. It's necessary for `Drop` in the Rust bindings, and being necessary for the Rust bindings is often a sign that it should be used in the C too, because Rust just enforces good C practices :)

I'm wiping here because the C can't guarantee lifetimes the way the Rust can, so this is a safety measure against misuse. If nothing else, we don't want the previous result to be accessible after a new ClientHello is set.

However, the application probably doesn't want to only rely on this safety measure. Look at this flow:
```
struct s2n_fingerprint *fingerprint = s2n_fingerprint_new();
for (size_t i = 0; i < 10; i++) {
    struct s2n_client_hello *hello = hello_from_some_bytes();
    s2n_fingerprint_set_client_hello(hello);
    // At this point, fingerprint has a reference to a valid s2n_client_hello
    s2n_fingerprint_get_hash(blah);
    s2n_client_hello_free(hello);
    // At this point, fingerprint has a reference to an invalid s2n_client_hello
    // It's probably fine if we intentionally don't use it, but it's still not great.
    // If we can call wipe first, it'd be better.
}
```

Reply:
Makes sense thanks!